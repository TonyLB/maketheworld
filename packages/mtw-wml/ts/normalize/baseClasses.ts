type TagType = 'Asset' |
    'Story' |
    'Character' |
    'Import' |
    'If' |
    'Description' |
    'Room' |
    'Feature' |
    'Image' |
    'Map' |
    'Exit' |
    'Variable' |
    'Computed' |
    'Action' |
    'Character'

export type NormalReference = {
    key: string;
    tag: TagType;
    index: number;
}

export type BaseAppearance = {
    contextStack: NormalReference[];
    contents: NormalReference[];
    location?: number[];
}

type NormalBase = {
    key: string;
}

export type NormalAsset = {
    tag: 'Asset';
    Story?: boolean;
    instance?: boolean;
    appearances: BaseAppearance[];
    fileName?: string;
    zone?: string;
} & NormalBase

export type NormalCharacterPronouns = {
    subject: string;
    object: string;
    reflexive: string;
    possessive: string;
    adjective: string;
}

export type NormalCharacter = {
    tag: 'Character';
    Name: string;
    Pronouns: NormalCharacterPronouns;
    FirstImpression: string;
    OneCoolThing: string;
    Outfit: string;
    fileName?: string;
    fileURL?: string;
    appearances: BaseAppearance[];
} & NormalBase

export type ComponentRenderItem = {
    tag: 'Link';
    targetTag: 'Feature' | 'Action';
    to: string;
    text: string;
} | {
    tag: 'String';
    value: string;
} | {
    tag: 'LineBreak';
} | {
    tag: 'Space';
} | {
    tag: 'Condition';
    if: string;
    dependencies: string[];
    contents: ComponentRenderItem[];
}

export type NormalDescriptionPayload = {
    type: 'Description';
    render?: ComponentRenderItem[];
}

export type NormalDescription = NormalDescriptionPayload & NormalBase

export type ComponentAppearance = Omit<NormalDescriptionPayload, 'type'> & BaseAppearance & {
    name?: ComponentRenderItem[];
    x?: number;
    y?: number;
}

type ComponentTypes = 'Room' | 'Feature'

export type NormalComponent = {
    tag: ComponentTypes;
    appearances: ComponentAppearance[];
    global?: boolean;
} & NormalBase

export type NormalRoom = NormalComponent & {
    tag: 'Room';
}

export type NormalFeature = NormalComponent & {
    tag: 'Feature';
}

type NormalImportMapping = {
    key: string;
    type: string;
}

export type NormalImport = {
    tag: 'Import';
    from: string;
    mapping: Record<string, NormalImportMapping>;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalCondition = {
    tag: 'If';
    if: string;
    dependencies: string[];
    appearances: BaseAppearance[];
} & NormalBase

export type ImageAppearance = {
    display?: 'replace'
} & BaseAppearance

export type NormalImage = {
    tag: 'Image';
    fileURL?: string;
    appearances: ImageAppearance[];
} & NormalBase

export type MapAppearanceRoom = {
    conditions: {
        if: string;
        dependencies: string[]
    }[];
    key: string;
    x: number;
    y: number;
    location: number[];
}

type MapAppearanceImage = string

export type MapAppearance = {
    name: ComponentRenderItem[];
    images: MapAppearanceImage[];
    rooms: MapAppearanceRoom[];
} & BaseAppearance

export type NormalMap = {
    tag: 'Map';
    appearances: MapAppearance[];
} & NormalBase

type ExitAppearance = {
    name?: string;
} & BaseAppearance

export type NormalExit = {
    tag: 'Exit';
    to: string;
    from: string;
    name?: string;
    appearances?: ExitAppearance[];
} & NormalBase

export type NormalVariable = {
    tag: 'Variable';
    default: string;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalComputed = {
    tag: 'Computed';
    src: string;
    dependencies: string[];
    appearances: BaseAppearance[];
} & NormalBase

export type NormalAction = {
    tag: 'Action';
    src: string;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalItem = NormalAsset |
    NormalComponent |
    NormalImport |
    NormalImage |
    NormalMap |
    NormalExit |
    NormalVariable |
    NormalComputed |
    NormalAction |
    NormalCondition |
    NormalCharacter

export type NormalForm = Record<string, NormalItem>

export class WMLNormalizeError extends Error {
    constructor(message: string) {
        super(`WMLNormalize: ${message}`)
        this.name = 'WMLNormalizeError'
    }
}

export class NormalizeTagMismatchError extends WMLNormalizeError {
    constructor(message: string) {
        super(message)
        this.name = 'TagMismatchError'
    }
}

export class NormalizeKeyMismatchError extends WMLNormalizeError {
    constructor(message: string) {
        super(message)
        this.name = 'KeyTagMismatchError'
    }
}

export const isNormalExit = (arg: NormalItem): arg is NormalExit => (arg?.tag === 'Exit')
export const isNormalImage = (arg: NormalItem): arg is NormalImage => (arg?.tag === 'Image')
export const isNormalComponent = (arg: NormalItem): arg is NormalComponent => (['Room', 'Feature'].includes(arg?.tag))
export const isNormalRoom = (arg: NormalItem): arg is NormalRoom => (arg?.tag === 'Room')
export const isNormalFeature = (arg: NormalItem): arg is NormalFeature => (arg?.tag === 'Feature')
export const isNormalMap = (arg: NormalItem): arg is NormalMap => (arg?.tag === 'Map')
export function isNormalAsset(arg: NormalItem): arg is NormalAsset {
    return arg?.tag === 'Asset'
}
export const isNormalCharacter = (arg: NormalItem): arg is NormalCharacter => (arg?.tag === 'Character')
export const isNormalImport = (arg: NormalItem): arg is NormalImport => (arg?.tag === 'Import')
export const isNormalCondition = (arg: NormalItem): arg is NormalCondition => (arg.tag === 'If')
export const isNormalVariable = (arg: NormalItem): arg is NormalVariable => (arg.tag === 'Variable')
export const isNormalComputed = (arg: NormalItem): arg is NormalComputed => (arg.tag === 'Computed')
export const isNormalAction = (arg: NormalItem): arg is NormalAction => (arg.tag === 'Action')
