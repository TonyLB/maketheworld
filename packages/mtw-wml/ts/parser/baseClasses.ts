import {
    Token,
    TokenExpressionValue,
    TokenKeyValue,
    TokenLiteralValue,
    ParseStackTokenEntry
} from './tokenizer/baseClasses'

type ParseTagBase = {
    startTagToken: number;
    endTagToken: number;
}

export type ParseLiteralLegalContents = ParseStringTag
type ParseValueBase = {
    value: string;
    contents: ParseLiteralLegalContents[];
} & ParseTagBase

export type ParseAssetLegalContents = ParseActionTag |
    ParseCommentTag |
    ParseComputedTag |
    ParseConditionTag |
    ParseExitTag |
    ParseFeatureTag |
    ParseImageTag |
    ParseImportTag |
    ParseMapTag |
    ParseRoomTag |
    ParseVariableTag |
    ParseWhitespaceTag

type ParseAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseAssetTag = {
    tag: 'Asset';
} & ParseAssetBase

export type ParseStoryTag = {
    tag: 'Story';
    instance: boolean;
} & ParseAssetBase

export type ParseNameTag = ParseTaggedMessageTag<"Name">

export type ParsePronounsTag = {
    tag: 'Pronouns';
    subject: string;
    object: string;
    possessive: string;
    reflexive: string;
    adjective: string;
} & ParseTagBase

export type ParseOneCoolThingTag = {
    tag: 'OneCoolThing';
} & ParseValueBase

export type ParseFirstImpressionTag = {
    tag: 'FirstImpression';
} & ParseValueBase

export type ParseOutfitTag = {
    tag: 'Outfit';
} & ParseValueBase

export type ParseImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
} & ParseTagBase

export type ParseCharacterLegalContents = ParseNameTag | ParsePronounsTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag | ParseImageTag
export type ParseCharacterTag = {
    tag: 'Character';
    key: string;
    player?: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    contents: ParseCharacterLegalContents[];
} & ParseTagBase

export type ParseVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
} & ParseTagBase

export type ParseDependencyTag = {
    tag: 'Depend';
    on: string;
} & ParseTagBase

export type ParseComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: ParseDependencyTag[];
} & ParseTagBase

export type ParseActionTag = {
    tag: 'Action';
    key: string;
    src: string;
} & ParseTagBase

export type ParseUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
} & ParseTagBase

export type ParseImportLegalContents = ParseUseTag
export type ParseImportTag = {
    tag: 'Import';
    from: string;
    contents: ParseImportLegalContents[];
} & ParseTagBase

export type ParseConditionLegalContextTag = 'Asset' | 'Description' | 'Room' | 'Feature' | 'Map'

export type ParseConditionTagAssetContext = {
    tag: 'If';
    contextTag: 'Asset';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseConditionTagDescriptionContext = {
    tag: 'If';
    contextTag: 'Description';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseConditionTagRoomContext = {
    tag: 'If';
    contextTag: 'Room';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseConditionTagFeatureContext = {
    tag: 'If';
    contextTag: 'Feature';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseConditionTagMapContext = {
    tag: 'If';
    contextTag: 'Map';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseConditionTypeFromContextTag<T extends ParseConditionLegalContextTag> =
    T extends 'Description'
        ? ParseConditionTagDescriptionContext
        : T extends 'Room'
            ? ParseConditionTagRoomContext
            : T extends 'Feature'
                ? ParseConditionTagFeatureContext
                : T extends 'Map'
                    ? ParseConditionTagMapContext
                    : ParseConditionTagAssetContext

export type ParseConditionTag = ParseConditionTagAssetContext | ParseConditionTagDescriptionContext | ParseConditionTagRoomContext | ParseConditionTagFeatureContext | ParseConditionTagMapContext

export const isLegalParseConditionContextTag = (value: string): value is ParseConditionLegalContextTag => (['Asset', 'Description', 'Room', 'Feature', 'Map'].includes(value))
export const isParseConditionTagAssetContext = (value: ParseConditionTag): value is ParseConditionTagAssetContext => (value.contextTag === 'Asset')
export const isParseConditionTagDescriptionContext = (value: ParseConditionTag): value is ParseConditionTagDescriptionContext => (value.contextTag === 'Description')
export const isParseConditionTagRoomContext = (value: ParseConditionTag): value is ParseConditionTagRoomContext => (value.contextTag === 'Room')
export const isParseConditionTagFeatureContext = (value: ParseConditionTag): value is ParseConditionTagFeatureContext => (value.contextTag === 'Feature')
export const isParseConditionTagMapContext = (value: ParseConditionTag): value is ParseConditionTagMapContext => (value.contextTag === 'Map')

export const parseDifferentiatingTags: Record<ParseConditionLegalContextTag,  ParseTag["tag"][]> = {
    Asset: ['Exit', 'Feature', 'Room', 'If', 'Image', 'Map'],
    Description: ['If', 'Space', 'String', 'Link', 'br', 'Whitespace'],
    Room: ['If', 'Description', 'Name', 'Exit'],
    Feature: ['If', 'Description', 'Name'],
    Map: ['If', 'Image', 'Room', 'Name', 'Exit']
}

export type ParseExitTag = {
    tag: 'Exit';
    key?: string;
    name: string;
    to?: string;
    from?: string;
    contents: ParseStringTag[];
} & ParseTagBase

export type ParseLinkLegalContents = ParseStringTag | ParseWhitespaceTag
export type ParseLinkTag = {
    tag: 'Link';
    to: string;
    contents: ParseLinkLegalContents[];
} & ParseTagBase

export type ParseTaggedMessageLegalContents = ParseWhitespaceTag | ParseStringTag | ParseLinkTag | ParseLineBreakTag | ParseSpacerTag | ParseConditionTagDescriptionContext

export type ParseTaggedMessageTag<T extends string> = {
    tag: T;
    display?: 'before' | 'after' | 'replace';
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseDescriptionTag = ParseTaggedMessageTag<"Description">

export type ParseLineBreakTag = {
    tag: 'br';
} & ParseTagBase

export type ParseSpacerTag = {
    tag: 'Space';
} & ParseTagBase

export type ParseRoomLegalContents = ParseDescriptionTag | ParseNameTag | ParseExitTag | ParseFeatureTag | ParseConditionTagRoomContext
export type ParseRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseFeatureLegalContents = ParseDescriptionTag | ParseNameTag | ParseConditionTagFeatureContext
export type ParseFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseMapLegalContents = ParseNameTag | ParseRoomTag | ParseImageTag | ParseExitTag | ParseConditionTagMapContext
export type ParseMapTag = {
    tag: 'Map';
    key: string;
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseStringTag = {
    tag: 'String';
    value: string;
} & ParseTagBase

export type ParseWhitespaceTag = {
    tag: 'Whitespace';
} & ParseTagBase

export type ParseCommentTag = {
    tag: 'Comment';
} & ParseTagBase

export type ParseTag = ParseAssetTag |
    ParseStoryTag |
    ParseNameTag |
    ParseCharacterTag |
    ParsePronounsTag |
    ParseOneCoolThingTag |
    ParseFirstImpressionTag |
    ParseOutfitTag |
    ParseImageTag |
    ParseVariableTag |
    ParseComputedTag |
    ParseDependencyTag |
    ParseActionTag |
    ParseUseTag |
    ParseImportTag |
    ParseConditionTag |
    ParseExitTag |
    ParseDescriptionTag |
    ParseLineBreakTag |
    ParseLinkTag |
    ParseRoomTag |
    ParseFeatureTag |
    ParseMapTag |
    ParseStringTag |
    ParseWhitespaceTag |
    ParseSpacerTag |
    ParseCommentTag

export type ParseLegalTag = ParseTag["tag"]

export const isParseLegalTag = (tag: string): tag is ParseLegalTag => ([
    'Asset',
    'Story',
    'Name',
    'Character',
    'Pronouns',
    'OneCoolThing',
    'FirstImpression',
    'Outfit',
    'Image',
    'Variable',
    'Computed',
    'Depend',
    'Action',
    'Use',
    'Import',
    'If',
    'Exit',
    'Description',
    'br',
    'Link',
    'Room',
    'Feature',
    'Map',
    'String',
    'Whitespace',
    'Space',
    'Comment'
].includes(tag))

export type ParseStackTagOpenPendingEntry = {
    type: 'TagOpenPending';
    tag: string;
    startTagToken: number;
}

export type ParseStackTagOpenEntry = {
    type: 'TagOpen';
    tag: ParseLegalTag;
    startTagToken: number;
    properties: Record<string, (TokenExpressionValue | TokenKeyValue | TokenLiteralValue | boolean)>;
}

export type ParseStackTagEntry<T extends ParseTag> = {
    type: 'Tag';
    tag: T;
}

export class ParseException extends Error {
    startToken: number
    endToken: number
    constructor(message: string, startToken: number, endToken: number) {
        super(message)
        this.name = 'ParseException'
        this.startToken = startToken
        this.endToken = endToken
    }
}

export const isParseTagDependency = (value: ParseTag): value is ParseDependencyTag => (value.tag === 'Depend')
export const isParseTagNesting = (value: ParseTag): value is (ParseRoomTag | ParseFeatureTag | ParseAssetTag | ParseStoryTag | ParseCharacterTag | ParseImportTag | ParseDescriptionTag | ParseConditionTag | ParseLinkTag | ParseMapTag | ParseExitTag | ParseNameTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag) => (
    ['Room', 'Feature', 'Asset', 'Story', 'Character', 'Import', 'Description', 'If', 'Link', 'Map', 'Exit', 'Name', 'FirstImpression', 'OneCoolThing', 'Outfit'].includes(value.tag)
)
export const isParseExit = (value: ParseTag): value is ParseExitTag => (value.tag === 'Exit')
export const isParseRoom = (value: ParseTag): value is ParseRoomTag => (value.tag === 'Room')
export const isParseString = (value: ParseTag): value is ParseStringTag => (value.tag === 'String')

export type ParseStackEntry = ParseStackTagOpenPendingEntry | ParseStackTagOpenEntry | ParseStackTagEntry<ParseTag> | ParseStackTokenEntry<Token>

export const isParseStackTagOpenEntry = (value: ParseStackEntry): value is ParseStackTagOpenEntry => (value.type === 'TagOpen')

export type ParseTagFactory<T extends ParseTag> = (value: { open: ParseStackTagOpenEntry, context: ParseStackTagOpenEntry[]; contents: ParseTag[], endTagToken: number }) => ParseStackTagEntry<T>

export const parseTagDefaultProps: Record<string, string> = {
    If: 'if'
}
