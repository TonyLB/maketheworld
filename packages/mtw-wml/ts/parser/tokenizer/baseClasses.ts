import SourceStream from "./sourceStream";

export type TokenBase = {
    source: string;
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue'
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    key: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenProperty = {
    type: 'Property'
} & TokenBase

export type TokenBeginTagOpen = {
    type: 'BeginTagOpen'
} & TokenBase

export type TokenBeginTagClose = {
    type: 'BeginTagClose';
} & TokenBase

export type TokenEndTag = {
    type: 'EndTag';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenBeginTagOpen |
    TokenBeginTagClose |
    TokenEndTag |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | TokenError | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
