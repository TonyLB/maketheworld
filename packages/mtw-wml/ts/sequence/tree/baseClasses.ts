export type GenericTreeNodeFiltered<F extends {}, N extends {}> = {
    data: F;
    children: GenericTree<N>;
}

export type GenericTreeNode<N extends {}> = GenericTreeNodeFiltered<N, N>

export type GenericTreeFiltered<F extends {}, N extends {}> = GenericTreeNodeFiltered<F, N>[]

export type GenericTree<N extends {}> = GenericTreeFiltered<N, N>

export enum GenericTreeDiffAction {
    Exclude,
    Context,
    Set,
    Add,
    Delete
}

export type GenericTreeDiffNode<N extends {}> = {
    data: N;
    children: GenericTreeDiff<N>;
    action: GenericTreeDiffAction;
}

export type GenericTreeDiff<N extends {}> = GenericTreeDiffNode<N>[]

export type SourceWrap<N extends {}> = {
    contents: N;
    source: string;
}
