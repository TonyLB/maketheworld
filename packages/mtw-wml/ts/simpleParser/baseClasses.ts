export enum ParsePropertyTypes {
    Key,
    Literal,
    Expression
}

type ParsePropertyBase = {
    key: string;
}

export type ParsePropertyKey = {
    type: ParsePropertyTypes.Key;
    value: string;
} & ParsePropertyBase

export type ParsePropertyLiteral = {
    type: ParsePropertyTypes.Literal;
    value: string;
} & ParsePropertyBase

export type ParsePropertyExpression = {
    type: ParsePropertyTypes.Expression;
    value: string;
} & ParsePropertyBase

export type ParseProperty = ParsePropertyKey | ParsePropertyLiteral | ParsePropertyExpression

export enum ParseTypes {
    Open,
    SelfClosure,
    Close,
    Text
}

export type ParseTagOpen = {
    type: ParseTypes.Open;
    tag: string;
    properties: ParseProperty[];
}

export type ParseTagSelfClosure = {
    type: ParseTypes.SelfClosure;
    tag: string;
    properties: ParseProperty[];
}

export type ParseTagClose = {
    type: ParseTypes.Close;
    tag: string;
}

export type ParseFreeText = {
    type: ParseTypes.Text;
    text: string;
}

export type ParseItem = ParseTagOpen | ParseTagSelfClosure | ParseTagClose | ParseFreeText
