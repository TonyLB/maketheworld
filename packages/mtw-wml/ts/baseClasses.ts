import { ParseTag } from "./parser/baseClasses";

type SchemaNestingBase<T extends SchemaTag> = {
    contents: T[];
}

export type LegalAssetContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag
export type LegalConditionContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag

type SchemaAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
} & SchemaNestingBase<LegalAssetContents>

export type SchemaAssetTag = {
    tag: 'Asset';
} & SchemaAssetBase

export type SchemaStoryTag = {
    tag: 'Story';
    instance: boolean;
} & SchemaAssetBase

export type SchemaImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
}

type SchemaPronouns = {
    subject: string;
    object: string;
    possessive: string;
    adjective: string;
    reflexive: string;
}

export type SchemaCharacterTag = {
    tag: 'Character';
    player?: string;
    name: string;
    pronouns: SchemaPronouns;
    firstImpression?: string;
    oneCoolThing?: string;
    outfit?: string;
    image?: SchemaImageTag;
}

export type SchemaVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
}

export type SchemaComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: string[];
}

export type SchemaActionTag = {
    tag: 'Action';
    src: string;
}

export type SchemaUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
}

type SchemaImportMapping = {
    key: string;
    type: 'Room' | 'Feature' | 'Variable' | 'Computed' | 'Action' | 'Map'
}

export type SchemaImportTag = {
    tag: 'Import';
    from: string;
    mapping: Record<string, SchemaImportMapping>;
}

//
// TODO:  Figure out how to define limits on schemaConditionTag contents
// without causing circular reference
//
export type SchemaConditionTag = {
    tag: 'Condition';
    if: string;
    dependencies: string[];
} & SchemaNestingBase<any>

export type SchemaExitTag = {
    tag: 'Exit';
    key: string;
    name: string;
    to: string;
    from: string;
}

export type SchemaLinkTag = {
    tag: 'Link';
    to: string;
} & SchemaNestingBase<SchemaStringTag>

export type SchemaDescriptionTag = {
    tag: 'Description';
    display: 'before' | 'after' | 'replace';
    spaceBefore: boolean;
    spaceAfter: boolean;
} & SchemaNestingBase<SchemaStringTag | SchemaLinkTag | SchemaLineBreakTag>

export type SchemaLineBreakTag = {
    tag: 'br';
}

export type SchemaNameTag = {
    tag: 'Name';
    name: string;
}

export type SchemaRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
} & SchemaNestingBase<SchemaDescriptionTag | SchemaExitTag | SchemaFeatureTag | SchemaNameTag>

export type SchemaFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
}& SchemaNestingBase<SchemaDescriptionTag | SchemaNameTag>

export type SchemaMapTag = {
    tag: 'Map';
    key: string;
} & SchemaNestingBase<SchemaExitTag | SchemaImageTag | SchemaRoomTag>

export type SchemaStringTag = {
    tag: 'String';
    value: string;
}

export type SchemaTag = SchemaAssetTag |
    SchemaStoryTag |
    SchemaCharacterTag |
    SchemaImageTag |
    SchemaVariableTag |
    SchemaComputedTag |
    SchemaActionTag |
    SchemaUseTag |
    SchemaImportTag |
    SchemaConditionTag |
    SchemaExitTag |
    SchemaDescriptionTag |
    SchemaLineBreakTag |
    SchemaLinkTag |
    SchemaNameTag |
    SchemaRoomTag |
    SchemaFeatureTag |
    SchemaMapTag |
    SchemaStringTag

export class SchemaException extends Error {
    parseTag: ParseTag;
    constructor(message: string, parseTag: ParseTag) {
        super(message)
        this.name = 'WMLSchemaException'
        this.parseTag = parseTag
    }
}

export const isSchemaString = (value: SchemaTag): value is SchemaStringTag => (value.tag === 'String')
