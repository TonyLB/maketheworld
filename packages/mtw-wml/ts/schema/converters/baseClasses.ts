import { SchemaTag } from "../baseClasses"
import { ParsePropertyTypes, ParseTagOpen, ParseTagSelfClosure } from "../../simpleParser/baseClasses"
import { GenericTree, GenericTreeNode } from "../../tree/baseClasses";

export type SchemaConverterArguments = {
    parseOpen: ParseTagOpen | ParseTagSelfClosure;
    contextStack: GenericTree<SchemaTag>;
}

export type SchemaInitialConverter = {
    (args: SchemaConverterArguments): SchemaTag
}

export type ValidationTemplateItem = {
    required?: boolean;
    type: ParsePropertyTypes;
}

export type ValidationTemplate = Record<string, ValidationTemplateItem>

export type ValidationTemplateRemap<V extends ValidationTemplate>  ={
    [K in keyof V]: V[K] extends { type: ParsePropertyTypes.Boolean } ? boolean : string
}

export type ValidationRequiredKeys<V extends ValidationTemplate> = {[K in keyof V]: V[K] extends { required: true } ? K : never}[keyof V]

export type ValidationTemplateOutput<V extends ValidationTemplate> = 
    Partial<ValidationTemplateRemap<V>> &
    Pick<ValidationTemplateRemap<V>, ValidationRequiredKeys<V>>

export type ConverterMapValidateProperties = {
    isValid: (item: SchemaTag) => boolean;
    branchTags: SchemaTag['tag'][];
    leafTags: SchemaTag['tag'][];
}

export type ConverterMapEntry = {
    initialize: SchemaInitialConverter;
    typeCheckContents?: (item: SchemaTag, contextStack: GenericTree<SchemaTag>) => boolean;
    validateContents?: ConverterMapValidateProperties;
    finalize?: (initialTag: SchemaTag, contents: GenericTree<SchemaTag>, contextStack: GenericTree<SchemaTag>) => GenericTreeNode<SchemaTag>;
    //
    // wrapper and aggregate are for tags that wrap their sibling groups together (e.g. If/ElseIf/Else). If wrapper is
    // present but aggregate is not then a wrapper is created. If aggregate is present then a previous wrapper must
    // already exist on contextStack, and the aggregate function is used to turn that wrapper object (together with
    // its current children) into the new wrapper object (with, presumably, updated children)
    //
    wrapper?: 'If';
    aggregate?: (previous: GenericTreeNode<SchemaTag>, node: GenericTreeNode<SchemaTag>) => GenericTreeNode<SchemaTag>;
}

export const isSchemaWrapper = (tag: SchemaTag['tag']): tag is 'If' => (['If'].includes(tag))

export type SchemaToWMLOptions = {
    indent: number;
    context: SchemaTag[];
    siblings?: GenericTree<SchemaTag>;
}

export enum PrintMapOptionsChange {
    Sibling,
    Indent
}
export type PrintMapOptionsFactory = {
    (action: PrintMapOptionsChange): (previous: SchemaToWMLOptions) => SchemaToWMLOptions;
}

export type PrintMapEntryArguments = {
    tag: GenericTreeNode<SchemaTag>;
    options: SchemaToWMLOptions;
    optionsFactory: PrintMapOptionsFactory;
    schemaToWML: PrintMapEntry;
}

export enum PrintMode {
    naive,
    nested,
    propertyNested
}

//
// prettyPrint functions will always return PrintMapResult. The following assumptions should apply:
//    - If printMode === naive then the output should be a single line without line breaks
//    - If printMode === nested then the output can be multiline, with the left-most elements
//      snug against the left margin (no global indent ... indent is applied recursively as each
//      parent element incorporates its children)
//    - If printMode === nested then the output can be either a single element or multiple
//      elements
//    - If printMode === propertyNested then the output can represent only a single element (and
//      its children)
//
export type PrintMapResult = {
    printMode: PrintMode;
    output: string;
}

export type PrintMapEntry = {
    (args: PrintMapEntryArguments): PrintMapResult[];
}
