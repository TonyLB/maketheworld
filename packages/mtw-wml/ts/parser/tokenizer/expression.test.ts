import SourceStream from './sourceStream'
import expressionValueTokenizer, { expressionTemplateStringSubTokenizer } from './expression'
import { TokenizeException } from './baseClasses'

describe('expressionTokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    describe('templateStringSubTokenizer', () => {
        it('should return undefined when no match', () => {
            const testStream = new SourceStream('Testing')
            expect(expressionTemplateStringSubTokenizer(testStream)).toBe(undefined)
        })
        it('should tokenize nested template strings', () => {
            const testStream = new SourceStream("`Test${value + `}`}`<Room key=(Test)>")
            expect(expressionTemplateStringSubTokenizer(testStream)).toMatchSnapshot()
        })
        it('should not break on escaped quotes', () => {
            const testStream = new SourceStream("`\\`Test\\``<Room key=(Test)>")
            expect(expressionTemplateStringSubTokenizer(testStream)).toMatchSnapshot()
        })
        it('should return error on unbounded string literal', () => {
            const testStream = new SourceStream("`Testing'")
            expect(() => {
                expressionTemplateStringSubTokenizer(testStream)
            }).toThrow(TokenizeException)
            expect(testStream.isEndOfSource).toBe(true)
        })
    })

    it('should return undefined when no match', () => {
        const testStream = new SourceStream('Testing')
        expect(expressionValueTokenizer(testStream)).toBe(undefined)
    })
    it('should properly nest string literals', () => {
        const testStream = new SourceStream(`{ return 'Test' + " value}" }<Room key=(Test) />`)
        expect(expressionValueTokenizer(testStream)).toMatchSnapshot()
    })
    it('should properly nest template literals', () => {
        const testStream = new SourceStream("{ return `Test: ${value}` }<Room key=(Test) />")
        expect(expressionValueTokenizer(testStream)).toMatchSnapshot()
    })
    it('should properly nest expressions', () => {
        const testStream = new SourceStream("{ if (value === 'A') { return 'Test' } else { return 'Fail' } }<Room key=(Test) />")
        expect(expressionValueTokenizer(testStream)).toMatchSnapshot()
    })
    
})