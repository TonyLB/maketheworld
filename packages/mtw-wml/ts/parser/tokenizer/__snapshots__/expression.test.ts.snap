// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expressionTokenizer should properly nest expressions 1`] = `
Object {
  "endIdx": 62,
  "startIdx": 0,
  "type": "ExpressionValue",
  "value": "{ if (value === 'A') { return 'Test' } else { return 'Fail' } ",
}
`;

exports[`expressionTokenizer should properly nest string literals 1`] = `
Object {
  "endIdx": 28,
  "startIdx": 0,
  "type": "ExpressionValue",
  "value": "{ return 'Test' + \\" value}\\" ",
}
`;

exports[`expressionTokenizer should properly nest template literals 1`] = `
Object {
  "endIdx": 26,
  "startIdx": 0,
  "type": "ExpressionValue",
  "value": "{ return \`Test: \${value}\` ",
}
`;

exports[`expressionTokenizer templateStringSubTokenizer should not break on escaped quotes 1`] = `
Object {
  "endIdx": 9,
  "startIdx": 0,
  "type": "TemplateString",
}
`;

exports[`expressionTokenizer templateStringSubTokenizer should return error on unbounded string literal 1`] = `
Object {
  "endIdx": 8,
  "message": "Unbounded template string",
  "startIdx": 0,
  "type": "Error",
}
`;

exports[`expressionTokenizer templateStringSubTokenizer should tokenize nested template strings 1`] = `
Object {
  "endIdx": 19,
  "startIdx": 0,
  "type": "TemplateString",
}
`;
